### DISCUSSION TITLE
Does foreach iterate over uncommitted ledger?

### DISCUSSION THREAD

**Discussion by divneil-intel** [microsoft/CCF#3926] [Q&A] [answered]
From the documentation of `foreach`, it seems like this lambda works for KVs in local ledger. Can you pls confirm

https://microsoft.github.io/CCF/release/1.x/build_apps/kv/kv_how_to.html
"
Values can only be retrieved directly `(kv::MapHandle::get())` for a given target key. However, it is sometimes necessary to access unknown keys, or to iterate through all Key-Value pairs.
"

**Comment by achamayou** [comment, marked as answer]
`foreach()`, like `get()` and `put()` operates on the current read version, which may or may not be committed by consensus. It is guaranteed that if the read version is rolled back, the subsequent transaction containing the `foreach` will also be rolled back. Or to put it another way, if the transaction containing the `foreach` commits, then the read version it resolved to must have also committed.

There is not a foreach equivalent to `get_globally_committed()` if that's what you mean, that would let a transaction iterate over a map at the latest know committed value on a node.

Note that using `get_globally_committed()` in the context of a write transaction is dangerous and will potentially lead to a serializability break. If mixed with `get()`, it will also break snapshot isolation, since you may end up reading from two different versions. It is only safe to use `get_globally_committed()` in very restricted contexts, such as a read-only handler, doing all its reads through `get_globally_committed()`.

**Comment by eddyashton** [comment]
@divneil-intel `foreach` will iterate over all of the same entries that could be accessed by `get`. That includes entries written by the current transaction, and those written and applied by previously completed transactions. It does not include any results from other transactions still in-progress.

Those entries may not yet be committed by consensus (replicated on a majority of nodes), but this transaction will be applied in-order (ie - if it eventually commits, that implies that all the things it read are also committed, and if any of the transactions it depends on are rolled back, it is also rolled back).

**Comment by divneil-intel** [comment]
Thanks @achamayou, @eddyashton for your replies. The explanation makes the usage scenarios clear.
